xds/internal/balancer/clusterresolver/resource_resolver_dns.go

```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package clusterresolver
+
+import (
+	"fmt"
+
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+var (
+	newDNS = func(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+		dnsB := resolver.Get("dns")
+		return dnsB.Build(target, cc, opts)
+	}
+)
+
+// dnsResolver watches updates for the given DNS hostname.
+//
+// It implements resolver.ClientConn interface to work with the DNS resolver.
+type dnsResolver struct {
+	target           string
+	topLevelResolver *resourceResolver
+	r                resolver.Resolver
+
+	addrs          []string
+	updateReceived bool
+}
+
+func newDNSResolver(target string, topLevelResolver *resourceResolver) *dnsResolver {
+	ret := &dnsResolver{
+		target:           target,
+		topLevelResolver: topLevelResolver,
+	}
+	r, err := newDNS(resolver.Target{Scheme: "dns", Endpoint: target}, ret, resolver.BuildOptions{})
+	if err != nil {
+		select {
+		case <-topLevelResolver.updateChannel:
+		default:
+		}
+		topLevelResolver.updateChannel <- &resourceUpdate{err: err}
+	}
+	ret.r = r
+	return ret
+}
+
+func (dr *dnsResolver) lastUpdate() (interface{}, bool) {
+	if !dr.updateReceived {
+		return nil, false
+	}
+	return dr.addrs, true
+}
+
+func (dr *dnsResolver) resolveNow() {
+	dr.r.ResolveNow(resolver.ResolveNowOptions{})
+}
+
+func (dr *dnsResolver) stop() {
+	dr.r.Close()
+}
+
+// dnsResolver needs to implement resolver.ClientConn interface to receive
```

Comment:
The methods implementing the `resolver.ClientConn` interface are not guarded by any mutex? Is that fine?

Also, there is no check if `dosnResolver.stop()` was invoked and therefore making these a no-op?